libs/vkd3d/command.c

```diff
@@ -1412,35 +1421,117 @@ static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_GetDevice(d3d12_fence_iface
+{
+    const struct vkd3d_vk_device_procs *vk_procs;
+    struct d3d12_shared_fence *fence;
+    VkSemaphoreWaitInfo wait_info;
     uint64_t completed_value;
     VkResult vr;
 
-    TRACE("iface %p\n", iface);
+    fence = userdata;
+    vk_procs = &fence->device->vk_procs;
+
+    wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
+    wait_info.pNext = NULL;
+    wait_info.flags = 0;
+    wait_info.semaphoreCount = 1;
+    wait_info.pSemaphores = &fence->timeline_semaphore;
+    wait_info.pValues = NULL;
 
-    vr = VK_CALL(vkGetSemaphoreCounterValueKHR(fence->device->vk_device, fence->timeline_semaphore, &completed_value));
-    if (vr != VK_SUCCESS)
+    while (!vkd3d_atomic_uint32_load_explicit(&fence->stop, vkd3d_memory_order_acquire))
     {
-        ERR("Failed to get shared fence counter value, error %d.\n", vr);
-        return 0;
+        vr = VK_CALL(vkGetSemaphoreCounterValueKHR(fence->device->vk_device, fence->timeline_semaphore, &completed_value));
+        if (vr != VK_SUCCESS)
+        {
+            ERR("Failed to get shared fence counter value, error %d.\n", vr);
+            return NULL;
+        }
+
+        vkd3d_shared_fence_signal_events(fence, completed_value);
+
+        if (!!vkd3d_atomic_uint32_load_explicit(&fence->stop, vkd3d_memory_order_acquire))
+            return NULL;
+
+        completed_value++;
+        wait_info.pValues = &completed_value;
+        vr = VK_CALL(vkWaitSemaphoresKHR(fence->device->vk_device, &wait_info, 10000000ull));
+        if (vr != VK_SUCCESS && vr != VK_TIMEOUT)
+        {
+            ERR("Failed to wait for semaphore, error %d.\n", vr);
+            return NULL;
+        }
     }
-    return completed_value;
+
+    return NULL;
 }
 
 static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_SetEventOnCompletion(d3d12_fence_iface *iface,
         UINT64 value, HANDLE event)
 {
     struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
     const struct vkd3d_vk_device_procs *vk_procs = &fence->device->vk_procs;
+    struct vkd3d_shared_fence_waiting_event *waiting_event;
     VkSemaphoreWaitInfo wait_info;
+    uint64_t completed_value;
     VkResult vr;
 
     TRACE("iface %p, value %#"PRIx64", event %p.\n", iface, value, event);
 
     if (event)
     {
-        FIXME("Signaling events on shared fence completion not supported.\n");
-        return E_NOTIMPL;
+        completed_value = d3d12_shared_fence_GetCompletedValue(iface);
+        if (completed_value >= value)
+        {
+            fence->device->signal_event(event);
+            return S_OK;
+        }
+
+        pthread_mutex_lock(&fence->mutex);
+
+        if (!(waiting_event = vkd3d_malloc(sizeof(*waiting_event))))
+        {
+            ERR("Failed to register device singleton for adapter.");
+            pthread_mutex_unlock(&fence->mutex);
+            return E_OUTOFMEMORY;
+        }
+        waiting_event->value = value;
+        waiting_event->event = event;
```

Comment:
Need to handle the scenario where event is NULL, in which case we just need to block in this call until it's done. Could just call WaitSemaphoresKHR directly here.

Reply:
That's exactly how this works at the moment. This is inside an `if(event)`, so if the event is null, we just go to the `vkWaitSemaphores` call at the end of the function.

Reply:
Oh, I see ... Ye, the early return was a bit surprising here.